xlators/storage/posix/src/posix-inode-fd-ops.c

```diff
@@ -2529,6 +2570,9 @@ posix_release(xlator_t *this, fd_t *fd)
     VALIDATE_OR_GOTO(this, out);
     VALIDATE_OR_GOTO(fd, out);
 
+    if (fd->inode && fd->inode->fd_count == 0)
+        posix_unlink_renamed_file_and_free_ctx(this, fd->inode);
```

Comment:
Is it really necessary to do this here ? once the file is closed and the reference to the inode removed, shouldn't `posix_forget()` take care of the inode cleanup ?

Reply:
Once the final reference to the file is closed/released, we receive a call to fd_destroy() ==> posix_release().
I have not observed posix_forget() being called when the file is closed. In case of unlink, we see posix_forget() is invoked and not seen posix_release() during unlink.

 I have captured call stack for both file close() and unlink

1. call stack for file close() ===>posix_release() is triggered

Thread 17 "glfs_rpcrqhnd" hit Breakpoint 1, posix_release (this=0x7fb4c000a140, fd=0x7fb49001aeb8) at posix-inode-fd-ops.c:2566
2566	    uint64_t tmp_pfd = 0;
(gdb) bt
#0  posix_release (this=0x7fb4c000a140, fd=0x7fb49001aeb8) at posix-inode-fd-ops.c:2566
#1  0x00007fb4d9a3ca05 in fd_destroy (fd=0x7fb49001aeb8) at fd.c:479
#2  fd_unref (fd=0x7fb49001aeb8) at fd.c:546
#3  0x00007fb4c702343b in server4_0_release (req=0x7fb4b8003488) at server-rpc-fops_v2.c:4042
#4  0x00007fb4d99ba025 in rpcsvc_request_handler (arg=0x7fb4c6f84140) at rpcsvc.c:2232
#5  0x00007fb4d97a4432 in start_thread () from /lib64/libpthread.so.0
#6  0x00007fb4d93e39d3 in clone () from /lib64/libc.so.6

2. call stack for unlink of a file() ===> posiz_forget() gets triggered
Thread 23 "glfs_iotwr006" hit Breakpoint 2, posix_forget (this=0x7fb4c000a140, inode=0x7fb4900144f8) at posix-inode-fd-ops.c:6005
6005	    return posix_unlink_renamed_file_and_free_ctx(this, inode);
(gdb) bt
#0  posix_forget (this=0x7fb4c000a140, inode=0x7fb4900144f8) at posix-inode-fd-ops.c:6005
#1  0x00007fb4d9a26b50 in __inode_ctx_free (inode=<optimized out>) at inode.c:348
#2  __inode_ctx_free (inode=0x7fb4900144f8) at inode.c:331
#3  0x00007fb4d9a28059 in __inode_destroy (inode=0x7fb4900144f8) at inode.c:1628
#4  inode_table_prune (table=table@entry=0x7fb4c007da50) at inode.c:1628
#5  0x00007fb4d9a281b8 in inode_unref (inode=0x7fb4900144f8) at inode.c:603
#6  0x00007fb4d9a15472 in loc_wipe (loc=loc@entry=0x7fb4b0027040) at xlator.c:1020
#7  0x00007fb4d9a8ea32 in args_wipe (args=0x7fb4b0027040) at default-args.c:1622
#8  0x00007fb4d9a3ba49 in call_stub_wipe_args (stub=0x7fb4b0026ff8) at call-stub.c:2367
#9  call_stub_destroy (stub=0x7fb4b0026ff8) at call-stub.c:2367
#10 0x00007fb4c7177740 in iot_worker (data=0x7fb4c0056570) at io-threads.c:232
#11 0x00007fb4d97a4432 in start_thread () from /lib64/libpthread.so.0
#12 0x00007fb4d93e39d3 in clone () from /lib64/libc.so.6

Reply:
The `fd` keeps a reference of the inode, so `posix_forget` cannot be called until `fd` has been completely destroyed. Even then, it might happen that other xlators keep a reference to the inode, so `posix_forget` may be called a bit later or even from another thread.

You should check if after continuing from `posix_release`, `posix_forget` is not called. If it's not called after some time, we might have an inode leak.

Reply:
Yeah, It works as expected. Removing the unnecessary code from posix_release()